source/volume_of_fluid/utilities.cc

```diff
@@ -0,0 +1,687 @@
+        if (vol<v2)
+          {
+            return 0.5*(-1+m1+sqrt(m1*m1+8*m2*m3*(vol-v1)));
+          }
+
+        // Case 3
+        double v3 = 0.0;
+        if (m12<m3)
+          {
+            v3 += m3*m3*(3*m12-m3)/(6.0*mprod);
+            v3 += m1*m1*(m1-3*m3)/(6*mprod);
+            v3 += m2*m2*(m2-3*m3)/(6*mprod);
+          }
+        else
+          {
+            v3 = 0.5*m12/m3;
+          }
+        if (vol<v3)
+          {
+            // Solve appropriate cubic
+            double a2 = -3.0*m12;
+            double a1 = 3.0*(m1*m1+m2*m2);
+            double a0 = 6*mprod*vol-m1*m1*m1-m2*m2*m2;
+            double np0 = a2*a2/9.0-a1/3.0;
+            double q0 = (a1*a2-3.0*a0)/6.0-a2*a2*a2/27.0;
+            double theta = acos(q0/sqrt(np0*np0*np0))/3.0;
+            return sqrt(np0)*(sqrt(3.0)*sin(theta)-cos(theta))-a2/3.0;
+          }
+
+        // Case 4
+        if (m3<m12)
+          {
+            // Solve appropriate cubic
+            double a2 = -1.5;
+            double a1 = 1.5*(m1*m1+m2*m2+m3*m3);
+            double a0 = 6*mprod*vol-m1*m1*m1-m2*m2*m2-m3*m3*m3;
+            double np0 = a2*a2/9.0-a1/3.0;
+            double q0 = (a1*a2-3.0*a0)/6.0-a2*a2*a2/27.0;
+            double theta = acos(q0/sqrt(np0*np0*np0))/3.0;
+            return sqrt(np0)*(sqrt(3.0)*sin(theta)-cos(theta))-a2/3.0;
+          }
+
+        return -0.5+m1*vol+0.5*m12;
+      }
+
+      template<>
+      void xFEM_Heaviside<2>(const int degree,
+                             const Tensor<1, 2> normal,
+                             const double d,
+                             const std::vector<Point<2>> &points,
+                             std::vector<double> &values)
+      {
+        const int basis_count=4;
+        std::vector<double> coeffs(basis_count);
+
+        const double n_xp = fabs(normal[0]), n_yp = fabs(normal[1]);
+        const double sign_n_x = (((normal[0]) > 0) - ((normal[0]) < 0)),
+                     sign_n_y = (((normal[1]) > 0) - ((normal[1]) < 0));
+
+        const double norm1 = n_xp + n_yp;
+        const double triangle_break = 0.5L*fabs(n_xp-n_yp);
+
+        const int max_degree = 1;
+
+        AssertThrow(degree<=max_degree,
+                    ExcMessage("Cannot generate xFEM polynomials are only functional for degrees<2."));
+
+        // Values calculated using sympy
+        if (d<-0.5*norm1)
+          {
+            for (int i =0; i < basis_count; ++i)
+              coeffs[i] = 0.0;
+          }
+        else if (d>0.5*norm1)
+          {
+            // Full cell
+            coeffs[0] = 1.0;
+            for (int i =1; i < basis_count; ++i)
+              coeffs[i] = 0.0;
+          }
+        else if (norm1< 1e-7)
+          {
+            coeffs[0] = 0.5;
+            for (int i =1; i < basis_count; ++i)
+              coeffs[i] = 0.0;
+          }
+        else if (d<=-triangle_break)
+          {
+            //Triangle
+            const double d_n = d + 0.5*norm1;
```

Comment:
In the formulas below, are the `L`'s necessary? if note remove. And replace the 2 and 3 by 2.0 and 3.0 to avoid confusion.